cobol-parser/src/main/scala/za/co/absa/cobrix/cobol/parser/CopybookParser.scala

```diff
@@ -162,15 +164,63 @@ object CopybookParser {
     })
     val schema: MutableCopybook = ArrayBuffer(root)
 
+    val nonTerms: Set[String] = (for (id <- nonTerminals)
+      yield transformIdentifier(id)
+    ).toSet
+
     val newTrees = if (dropGroupFillers) {
-      calculateNonFillerSizes(markSegmentRedefines(processGroupFillers(markDependeeFields(calculateBinaryProperties(schema))), segmentRedefines))
+      calculateNonFillerSizes(markSegmentRedefines(processGroupFillers(markDependeeFields(
+        addNonTerminals(calculateBinaryProperties(schema), nonTerms, enc, stringTrimmingPolicy, ebcdicCodePage)
+      )), segmentRedefines))
     } else {
-      calculateNonFillerSizes(markSegmentRedefines(renameGroupFillers(markDependeeFields(calculateBinaryProperties(schema))), segmentRedefines))
+      calculateNonFillerSizes(markSegmentRedefines(renameGroupFillers(markDependeeFields(
+        addNonTerminals(calculateBinaryProperties(schema), nonTerms, enc, stringTrimmingPolicy, ebcdicCodePage)
+      )), segmentRedefines))
     }
 
     new Copybook(newTrees.head.asInstanceOf[Group])
   }
 
+  private def addNonTerminals(copybook: MutableCopybook, nonTerminals: Set[String],
+                              enc: Encoding,
+                              stringTrimmingPolicy: StringTrimmingPolicy,
+                              ebcdicCodePage: CodePage
+                             ): MutableCopybook = {
+    val newCopybook: MutableCopybook = new ArrayBuffer()
+    for(stmt <- copybook) {
+      stmt match {
+        case s: Primitive => newCopybook.append(s)
+        case g: Group => {
+          if (nonTerminals contains g.name) {
+            newCopybook.append(
+              g.copy(
+                children = addNonTerminals(g.children, nonTerminals, enc, stringTrimmingPolicy, ebcdicCodePage),
+                isRedefined = true
+              )(g.parent)
+            )
+            val sz = g.binaryProperties.actualSize
+            val dataType = AlphaNumeric(s"X($sz)", sz, enc = Some(enc))
+            val decode = DecoderSelector.getDecoder(dataType, stringTrimmingPolicy, ebcdicCodePage)
+            newCopybook.append(
+              Primitive(
+                g.level, g.name + "_NT", g.lineNumber,
```

Comment:
* The name generator for non-terminals can be extracted as a function.
* The function can also check if the generated name does not exist.
* The "_NT" postfix can be moved to Constants in case we might need it in a different context.

Reply:
Added a new function to come up with new names in case of clashes at the parent level.